File: 1715-split-a-string-into-the-max-number-of-unique-substrings/split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        result = 0
        
        def dfs(i, curr):
            if i == len(s):
                return 0

            result = 0
            for j in range(i + 1, len(s) + 1):
                if s[i:j] in curr:
                    continue
                
                curr.add(s[i:j])
                result = max(result, 1 + dfs(j, curr))
                curr.remove(s[i:j])
            return result
            
        return dfs(0, set())

File: 1715-split-a-string-into-the-max-number-of-unique-substrings/split-a-string-into-the-max-number-of-unique-substrings.java
class Solution {
    int dfs(int i, HashSet<String> cache, String s){
        if (i == s.length()) return 0;
        int result = 0;
        for (int j = i + 1; j <= s.length(); j++){
            String substring = s.substring(i, j);
            if (cache.contains(substring)) continue;
            cache.add(substring);
            result = Math.max(result, 1 + dfs(j, cache, s));
            cache.remove(substring);
        }
        return result;
    }

    public int maxUniqueSplit(String s) {
        HashSet<String> cache = new HashSet<String>();
        return dfs(0, cache, s);        
    }
}